simple_parsing/annotation_utils/get_field_annotations.py

```diff
@@ -178,9 +177,17 @@ def get_field_type_from_annotations(some_class: type, field_name: str) -> type:
     # Get the local and global namespaces to pass to the `get_type_hints` function.
     local_ns: Dict[str, Any] = {"typing": typing, **vars(typing)}
     local_ns.update(forward_refs_to_types)
-    # Get the globals in the module where the class was defined.
-    global_ns = sys.modules[some_class.__module__].__dict__
-
+    # Get the globals in the module where the class and the bases of the class were defined, recursively.
+    global_ns = {}
+    
+    def add_global_ns(some_cls: type):
+        nonlocal global_ns
+        global_ns.update(sys.modules[some_cls.__module__].__dict__)
+        for base_cls in some_cls.__bases__:
+            add_global_ns(base_cls)
+
+    add_global_ns(some_class)
```

Comment:
Correct me if I'm wrong, but I don't think the function, nonlocal, or recursion are needed here:

```suggestion
    global_ns = sys.modules[some_cls.__module__].__dict__
    for base_cls in some_cls.__bases__:
        global_ns.update(sys.modules[base_cls.__module__].__dict__)
```

Reply:
You are right, @lebrice.
I've double checked the __bases__ attribute. It contains the list of all classes the given class inherits.

Reply:
The recursive get_global_ns will pass the above test.
I am guessing the Fingers in ```FieldComponents("Fingers", "1", 1, "foobar")``` is evaluated afterward and it needs to be evaluated recursively.

We will have to keep ```add_global_ns``` to pass all tests.

Reply:
Let's look at following captured logs for test ```test/test_literal.py::test_literal[verbose-literal_field9]```.
By executing the following code:
```Python
global_ns = sys.modules[some_class.__module__].__dict__
    print(some_class,some_class.__module__)
    for base_cls in some_class.__bases__:
        print(base_cls,base_cls.__module__)
        global_ns.update(sys.modules[base_cls.__module__].__dict__)
```
It failed and prints:
```Python
<class 'test.testutils.test_literal.<locals>.Foo'> test.testutils
<class 'test.testutils.TestSetup'> test.testutils
```

By executing the following code:
```Python
    def add_global_ns(some_cls: type):
        nonlocal global_ns
        global_ns.update(sys.modules[some_cls.__module__].__dict__)
        print(some_cls,some_cls.__module__)
        for base_cls in some_cls.__bases__:
            add_global_ns(base_cls)

    add_global_ns(some_class)
```
It passed and prints:
```Python
<class 'test.test_literal.test_literal.<locals>.Foo'> test.test_literal
<class 'test.testutils.TestSetup'> test.testutils
<class 'object'> builtins
```

The local defined Foo is bounded to a different module in the run time.
That makes us have to evaluate the Foo recursively.
How do you think, @lebrice ?

Reply:
I don't understand the difference. I'll have to look into this.

Reply:
There's also a problem with this implementation, which I mentioned in #168: The namespace of the base classes overwrites the global namespace of the subclass. This is a corner case, and I'm not 100% sure how to solve it.
Naively not overwriting the dictionary entries would cause subtle issues. 
```python
# base.py
from __future__ import annotations
@dataclass
class Foo:
    ...

@dataclass
class Base:
    a: Foo
```
```python
#subclass.py
from __future__ import annotations
from .base import Base
@dataclass
class Foo:
    something_else: bool = True
    
@dataclass
class Subclass(Base):
    other_attribute: Foo
```

With the current code, parsing an instance of `Subclass` would make both attributes have a `base.Foo` attribute, instead of having the right type for each field.

I don't currently have any idea on how to properly solve this, but I'm sure it exists.
You probably don't need to fix this in this PR, but I just wanted you to know about it.

Reply:
It will take me some time to think about that.
I just found some references that might be helpful to read: [postponed annotation in pydantic](https://github.com/pydantic/pydantic/blob/e34ff9255f281eaccac10d0436c082626d9146c3/docs/usage/postponed_annotations.md)
Here is their implementation: [src](https://github.com/pydantic/pydantic/blob/a7656fcd0965e2dfdabbbbe72309dc5211e6115c/pydantic/_internal/_typing_extra.py#L247).